Index: linux-2.6.28-rc6/arch/arm/mach-davinci/include/mach/irqs.h
===================================================================
--- linux-2.6.28-rc6.orig/arch/arm/mach-davinci/include/mach/irqs.h
+++ linux-2.6.28-rc6/arch/arm/mach-davinci/include/mach/irqs.h
@@ -95,8 +95,16 @@
 #define IRQ_COMMRX       62
 #define IRQ_EMUINT       63
 
+#ifdef CONFIG_ARCH_OMAPL1x7
+#define DAVINCI_N_AINTC_IRQ	91
+#else
 #define DAVINCI_N_AINTC_IRQ	64
+#endif
+#ifdef CONFIG_ARCH_OMAPL1x7
+#define DAVINCI_N_GPIO		128
+#else
 #define DAVINCI_N_GPIO		104
+#endif
 
 #define NR_IRQS			(DAVINCI_N_AINTC_IRQ + DAVINCI_N_GPIO)
 
@@ -203,4 +211,101 @@
 #define IRQ_DM355_GPIOBNK5	59
 #define IRQ_DM355_GPIOBNK6	60
 
+/* OMAPL1X7 interrupts */
+#define IRQ_OMAPL1X7_COMMTX                0
+#define IRQ_OMAPL1X7_COMMRX                1
+#define IRQ_OMAPL1X7_NINT                  2
+#define IRQ_OMAPL1X7_EVTOUT0               3
+#define IRQ_OMAPL1X7_EVTOUT1               4
+#define IRQ_OMAPL1X7_EVTOUT2               5
+#define IRQ_OMAPL1X7_EVTOUT3               6
+#define IRQ_OMAPL1X7_EVTOUT4               7
+#define IRQ_OMAPL1X7_EVTOUT5               8
+#define IRQ_OMAPL1X7_EVTOUT6               9
+#define IRQ_OMAPL1X7_EVTOUT7               10
+#define IRQ_OMAPL1X7_CCINT0                11
+#define IRQ_OMAPL1X7_CCERRINT              12
+#define IRQ_OMAPL1X7_TCERRINT0             13
+#define IRQ_OMAPL1X7_AEMIFINT              14
+#define IRQ_OMAPL1X7_I2CINT0               15
+#define IRQ_OMAPL1X7_MMCSDINT0             16
+#define IRQ_OMAPL1X7_MMCSDINT1             17
+#define IRQ_OMAPL1X7_ALLINT0               18
+#define IRQ_OMAPL1X7_RTC                   19
+#define IRQ_OMAPL1X7_SPINT0                20
+#define IRQ_OMAPL1X7_TINT12_0              21
+#define IRQ_OMAPL1X7_TINT34_0              22
+#define IRQ_OMAPL1X7_TINT12_1              23
+#define IRQ_OMAPL1X7_TINT34_1              24
+#define IRQ_OMAPL1X7_UARTINT0              25
+#define IRQ_OMAPL1X7_KEYMGRINT             26
+#define IRQ_OMAPL1X7_SECINT                26
+#define IRQ_OMAPL1X7_SECKEYERR             26
+#define IRQ_OMAPL1X7_MPUERR                27
+#define IRQ_OMAPL1X7_IOPUERR               27
+#define IRQ_OMAPL1X7_BOOTCFGERR            27
+#define IRQ_OMAPL1X7_CHIPINT0              28
+#define IRQ_OMAPL1X7_CHIPINT1              29
+#define IRQ_OMAPL1X7_CHIPINT2              30
+#define IRQ_OMAPL1X7_CHIPINT3              31
+#define IRQ_OMAPL1X7_TCERRINT1             32
+#define IRQ_OMAPL1X7_C0_RX_THRESH_PULSE    33
+#define IRQ_OMAPL1X7_C0_RX_PULSE           34
+#define IRQ_OMAPL1X7_C0_TX_PULSE           35
+#define IRQ_OMAPL1X7_C0_MISC_PULSE         36
+#define IRQ_OMAPL1X7_C1_RX_THRESH_PULSE    37
+#define IRQ_OMAPL1X7_C1_RX_PULSE           38
+#define IRQ_OMAPL1X7_C1_TX_PULSE           39
+#define IRQ_OMAPL1X7_C1_MISC_PULSE         40
+#define IRQ_OMAPL1X7_MEMERR                41
+#define IRQ_OMAPL1X7_GPIO0                 42
+#define IRQ_OMAPL1X7_GPIO1                 43
+#define IRQ_OMAPL1X7_GPIO2                 44
+#define IRQ_OMAPL1X7_GPIO3                 45
+#define IRQ_OMAPL1X7_GPIO4                 46
+#define IRQ_OMAPL1X7_GPIO5                 47
+#define IRQ_OMAPL1X7_GPIO6                 48
+#define IRQ_OMAPL1X7_GPIO7                 49
+#define IRQ_OMAPL1X7_GPIO8                 50
+#define IRQ_OMAPL1X7_I2CINT1               51
+#define IRQ_OMAPL1X7_LCDINT                52
+#define IRQ_OMAPL1X7_UARTINT1              53
+#define IRQ_OMAPL1X7_MCASPINT              54
+#define IRQ_OMAPL1X7_ALLINT1               55
+#define IRQ_OMAPL1X7_SPINT1                56
+#define IRQ_OMAPL1X7_UHPI_INT1             57
+#define IRQ_OMAPL1X7_USB_INT               58
+#define IRQ_OMAPL1X7_IRQN                  59
+#define IRQ_OMAPL1X7_RWAKEUP               60
+#define IRQ_OMAPL1X7_UARTINT2              61
+#define IRQ_OMAPL1X7_DFTSSINT              62
+#define IRQ_OMAPL1X7_EHRPWM0               63
+#define IRQ_OMAPL1X7_EHRPWM0TZ             64
+#define IRQ_OMAPL1X7_EHRPWM1               65
+#define IRQ_OMAPL1X7_EHRPWM1TZ             66
+#define IRQ_OMAPL1X7_EHRPWM2               67
+#define IRQ_OMAPL1X7_EHRPWM2TZ             68
+#define IRQ_OMAPL1X7_ECAP0                 69
+#define IRQ_OMAPL1X7_ECAP1                 70
+#define IRQ_OMAPL1X7_ECAP2                 71
+#define IRQ_OMAPL1X7_EQEP0                 72
+#define IRQ_OMAPL1X7_EQEP1                 73
+#define IRQ_OMAPL1X7_T12CMPINT0_0          74
+#define IRQ_OMAPL1X7_T12CMPINT1_0          75
+#define IRQ_OMAPL1X7_T12CMPINT2_0          76
+#define IRQ_OMAPL1X7_T12CMPINT3_0          77
+#define IRQ_OMAPL1X7_T12CMPINT4_0          78
+#define IRQ_OMAPL1X7_T12CMPINT5_0          79
+#define IRQ_OMAPL1X7_T12CMPINT6_0          80
+#define IRQ_OMAPL1X7_T12CMPINT7_0          81
+#define IRQ_OMAPL1X7_T12CMPINT0_1          82
+#define IRQ_OMAPL1X7_T12CMPINT1_1          83
+#define IRQ_OMAPL1X7_T12CMPINT2_1          84
+#define IRQ_OMAPL1X7_T12CMPINT3_1          85
+#define IRQ_OMAPL1X7_T12CMPINT4_1          86
+#define IRQ_OMAPL1X7_T12CMPINT5_1          87
+#define IRQ_OMAPL1X7_T12CMPINT6_1          88
+#define IRQ_OMAPL1X7_T12CMPINT7_1          89
+#define IRQ_OMAPL1X7_ARMCLKSTOPREQ         90
+
 #endif /* __ASM_ARCH_IRQS_H */
