Index: davinci-kernel/arch/arm/mach-davinci/id.c
===================================================================
--- davinci-kernel.orig/arch/arm/mach-davinci/id.c
+++ davinci-kernel/arch/arm/mach-davinci/id.c
@@ -15,7 +15,11 @@
 #include <linux/init.h>
 #include <linux/io.h>
 
+#ifdef CONFIG_ARCH_OMAPL1x7
+#define JTAG_ID_BASE		IO_ADDRESS(0x01c14018)
+#else
 #define JTAG_ID_BASE		IO_ADDRESS(0x01c40028)
+#endif
 
 static unsigned int davinci_revision;
 
@@ -49,6 +53,13 @@ static struct davinci_id davinci_ids[] _
 		.manufacturer	= 0x00f,
 		.type		= 0x03550000,
 	},
+	{
+		/* OMAPL1X7 */
+		.part_no	= 0xb7df,
+		.variant	= 0x0,
+		.manufacturer	= 0x017,
+		.type		= 0x08000000,
+	},
 };
 
 /*
@@ -111,6 +122,11 @@ void __init davinci_check_revision(void)
 		}
 	}
 
+#ifndef CONFIG_ARCH_OMAPL1x7
 	printk(KERN_INFO "DaVinci DM%04x variant 0x%x\n",
 	       davinci_rev(), variant);
+#else
+	printk(KERN_INFO "OMAP L1%04x variant 0x%x\n",
+	       davinci_rev(), variant);
+#endif
 }
Index: davinci-kernel/arch/arm/mach-davinci/include/mach/cpu.h
===================================================================
--- davinci-kernel.orig/arch/arm/mach-davinci/include/mach/cpu.h
+++ davinci-kernel/arch/arm/mach-davinci/include/mach/cpu.h
@@ -24,9 +24,16 @@ static inline int is_davinci_dm ##type(v
 	return (davinci_rev() == (id)) ? 1 : 0;	        \
 }
 
+#define IS_OMAP_CPU(type, id)			\
+static inline int is_omap ##type(void)	        \
+{							\
+	return (davinci_rev() == (id)) ? 1 : 0;	        \
+}
+
 IS_DAVINCI_CPU(644x, 0x6446)
 IS_DAVINCI_CPU(646x, 0x6467)
 IS_DAVINCI_CPU(355, 0x355)
+IS_OMAP_CPU(l1x7, 0x800)
 
 #ifdef CONFIG_ARCH_DAVINCI_DM644x
 #define cpu_is_davinci_dm644x() is_davinci_dm644x()
@@ -46,4 +53,10 @@ IS_DAVINCI_CPU(355, 0x355)
 #define cpu_is_davinci_dm355() 0
 #endif
 
+#ifdef CONFIG_ARCH_OMAPL1x7
+#define cpu_is_omapl1x7() is_omapl1x7()
+#else
+#define cpu_is_omapl1x7() 0
+#endif
+
 #endif
