Index: primus-kernel/arch/arm/mach-davinci/include/mach/mux.h
===================================================================
--- primus-kernel.orig/arch/arm/mach-davinci/include/mach/mux.h
+++ primus-kernel/arch/arm/mach-davinci/include/mach/mux.h
@@ -34,7 +34,7 @@
 struct mux_config {
 	char *name;
 	const char *mux_reg_name;
-	const unsigned char mux_reg;
+	const unsigned int mux_reg;
 	const unsigned char mask_offset;
 	const unsigned char mask;
 	const unsigned char mode;
@@ -151,6 +151,22 @@ enum davinci_dm355_index {
 	DM355_EVT26_MMC0_RX,
 };
 
+enum davinci_omapl1x7_index {
+	/* UART0 function */
+	OMAPL1X7_NUART0_CTS,
+	OMAPL1X7_NUART0_RTS,
+	OMAPL1X7_UART0_RXD,
+	OMAPL1X7_UART0_TXD,
+
+	/* UART1 function */
+	OMAPL1X7_UART1_RXD,
+	OMAPL1X7_UART1_TXD,
+
+	/* UART2 function */
+	OMAPL1X7_UART2_RXD,
+	OMAPL1X7_UART2_TXD,
+};
+
 #ifdef CONFIG_DAVINCI_MUX
 /* setup pin muxing */
 extern void davinci_mux_init(void);
Index: primus-kernel/arch/arm/mach-davinci/mux.h
===================================================================
--- primus-kernel.orig/arch/arm/mach-davinci/mux.h
+++ primus-kernel/arch/arm/mach-davinci/mux.h
@@ -48,4 +48,15 @@
 			.mode = mux_mode,				\
 		},
 
+#define OMAPL1X7_MUX_CFG(soc, desc, muxreg, mode_offset, mode_mask, mux_mode, dbg)\
+[soc##_##desc] = {							\
+			.name =  #desc,					\
+			.debug = dbg,					\
+			.mux_reg_name = "OMAPL1X7_PINMUX"#muxreg,	\
+			.mux_reg = OMAPL1X7_PINMUX##muxreg,		\
+			.mask_offset = mode_offset,			\
+			.mask = mode_mask,				\
+			.mode = mux_mode,				\
+		},
+
 #endif /* _MACH_DAVINCI_MUX_H */
Index: primus-kernel/arch/arm/mach-davinci/omapl1x7.c
===================================================================
--- primus-kernel.orig/arch/arm/mach-davinci/omapl1x7.c
+++ primus-kernel/arch/arm/mach-davinci/omapl1x7.c
@@ -160,7 +160,24 @@ static struct clk *omapl1x7_clks[] __ini
 	NULL,
 };
 
+/*
+ *		  soc		description	mux  mode   mode  mux	 dbg
+ *						reg  offset mask  mode
+ */
+static const struct mux_config davinci_omapl1x7_pins[] = {
+OMAPL1X7_MUX_CFG(OMAPL1X7,	NUART0_CTS,	7,   24,    15,	  2,	 false)
+OMAPL1X7_MUX_CFG(OMAPL1X7,	NUART0_RTS,	7,   28,    15,	  2,	 false)
+OMAPL1X7_MUX_CFG(OMAPL1X7,	UART0_RXD,	8,   12,    15,	  1,	 false)
+OMAPL1X7_MUX_CFG(OMAPL1X7,	UART0_TXD,	8,   16,    15,	  1,	 false)
+OMAPL1X7_MUX_CFG(OMAPL1X7,	UART1_RXD,	11,  8,     15,	  1,	 false)
+OMAPL1X7_MUX_CFG(OMAPL1X7,	UART1_TXD,	11,  12,    15,	  1,	 false)
+OMAPL1X7_MUX_CFG(OMAPL1X7,	UART2_RXD,	8,   28,    15,	  2,	 false)
+OMAPL1X7_MUX_CFG(OMAPL1X7,	UART2_TXD,	9,   0,     15,	  2,	 false)
+};
+
 void __init omapl1x7_init(void)
 {
 	davinci_clk_init(omapl1x7_clks);
+	davinci_mux_register(davinci_omapl1x7_pins,
+				ARRAY_SIZE(davinci_omapl1x7_pins));
 }
Index: primus-kernel/arch/arm/mach-davinci/mux.c
===================================================================
--- primus-kernel.orig/arch/arm/mach-davinci/mux.c
+++ primus-kernel/arch/arm/mach-davinci/mux.c
@@ -20,6 +20,8 @@
 #include <linux/spinlock.h>
 
 #include <mach/hardware.h>
+#include <mach/omapl1x7.h>
+#include <mach/cpu.h>
 #include <mach/mux.h>
 
 static const struct mux_config *mux_table;
@@ -49,6 +51,9 @@ int __init_or_module davinci_cfg_reg(con
 	if (!mux_table)
 		BUG();
 
+	if (cpu_is_omapl1x7())
+		base = IO_ADDRESS(OMAPL1X7_BOOT_CFG_BASE);
+
 	if (index >= pin_table_sz) {
 		printk(KERN_ERR "Invalid pin mux index: %lu (%lu)\n",
 		       index, pin_table_sz);
