Index: linux-2.6.28/arch/arm/mach-davinci/devices.c
===================================================================
--- linux-2.6.28.orig/arch/arm/mach-davinci/devices.c
+++ linux-2.6.28/arch/arm/mach-davinci/devices.c
@@ -463,6 +463,9 @@ static struct resource omapl1x7_emac_res
                 .name   = "mac_misc",
         },
 };
+static struct emac_platform_data omapl1x7_emac_pdata = {
+	.phy_mask = 0x0,	/* No PHY */
+};
 
 static struct platform_device omapl1x7_emac_device = {
         .name = "davinci_emac",
@@ -500,7 +503,7 @@ static struct platform_device omapl1x7_e
         .num_resources = ARRAY_SIZE(omapl1x7_emac_resources),
         .resource = omapl1x7_emac_resources,
         .dev = {
-                .platform_data = &emac_pdata,
+                .platform_data = &omapl1x7_emac_pdata,
         }
 };
 
Index: linux-2.6.28/arch/arm/mach-davinci/include/mach/emac.h
===================================================================
--- linux-2.6.28.orig/arch/arm/mach-davinci/include/mach/emac.h
+++ linux-2.6.28/arch/arm/mach-davinci/include/mach/emac.h
@@ -13,6 +13,7 @@
 
 struct emac_platform_data {
 	char mac_addr[6];
+	u32 phy_mask;
 };
 
 void davinci_init_emac(char *mac_addr);
