Index: linux-2.6.28/drivers/net/davinci_emac.c
===================================================================
--- linux-2.6.28.orig/drivers/net/davinci_emac.c
+++ linux-2.6.28/drivers/net/davinci_emac.c
@@ -200,6 +200,8 @@ static const char emac_version_string[] 
 #define EMAC_MACCONTROL_GIGABITEN_SHIFT (7)
 #define EMAC_MACCONTROL_FULLDUPLEXEN	(0x1)
 
+#define EMAC_MACCONTROL_RMIISPEED_MASK	(0x1 << 15)
+
 /* GIGABIT MODE related bits */
 #define EMAC_DM646X_MACCONTORL_GMIIEN	(0x01 << 5)
 #define EMAC_DM646X_MACCONTORL_GIG	(0x01 << 7)
@@ -1004,7 +1006,7 @@ static void emac_dev_mcast_set(struct ne
  */
 static void emac_int_disable(struct emac_priv *priv)
 {
-	if (cpu_is_davinci_dm646x()) {
+	if (cpu_is_davinci_dm646x() || cpu_is_omapl1x7()) {
 		unsigned long flags;
 
 		local_irq_save(flags);
@@ -1032,7 +1034,7 @@ static void emac_int_disable(struct emac
  */
 static void emac_int_enable(struct emac_priv *priv)
 {
-	if (cpu_is_davinci_dm646x()) {
+	if (cpu_is_davinci_dm646x() || cpu_is_omapl1x7()) {
 		emac_ctrl_write(EMAC_DM646X_CMRXINTEN, 0xff);
 		emac_ctrl_write(EMAC_DM646X_CMTXINTEN, 0xff);
 
@@ -2070,6 +2072,10 @@ static int emac_hw_enable(struct emac_pr
 		((priv->speed == 1000) ? EMAC_MACCONTROL_GIGABITEN : 0x0) |
 		((EMAC_DEF_TXPACING_EN) ? (EMAC_MACCONTROL_TXPACEEN) : 0x0) |
 		((priv->duplex == DUPLEX_FULL) ? 0x1 : 0));
+
+	if (cpu_is_omapl1x7())
+		mac_control |= EMAC_MACCONTROL_RMIISPEED_MASK;
+ 
 	emac_write(EMAC_MACCONTROL, mac_control);
 
 	mbp_enable =
@@ -2162,7 +2168,7 @@ static int emac_poll(struct napi_struct 
 	/* Since we support only 1 TX ch, for now check all TX int mask */
 	mask = EMAC_DM644X_MAC_IN_VECTOR_TX_INT_VEC;
 
-	if (cpu_is_davinci_dm646x())
+	if (cpu_is_davinci_dm646x() || cpu_is_omapl1x7())
 		mask = EMAC_DM646X_MAC_IN_VECTOR_TX_INT_VEC;
 
 	if (status & mask) {
@@ -2174,7 +2180,7 @@ static int emac_poll(struct napi_struct 
 	/* Since we support only 1 RX ch, for now check all RX int mask */
 	mask = EMAC_DM644X_MAC_IN_VECTOR_RX_INT_VEC;
 
-	if (cpu_is_davinci_dm646x())
+	if (cpu_is_davinci_dm646x() || cpu_is_omapl1x7())
 		mask = EMAC_DM646X_MAC_IN_VECTOR_RX_INT_VEC;
 
 	if (status & mask) {
