Index: primus-kernel/arch/arm/mach-davinci/gpio.c
===================================================================
--- primus-kernel.orig/arch/arm/mach-davinci/gpio.c
+++ primus-kernel/arch/arm/mach-davinci/gpio.c
@@ -130,6 +130,8 @@ static int __init davinci_gpio_setup(voi
 		 * are usable, and we're not rejecting the others.
 		 */
 		ngpio = 43;
+	} else if (cpu_is_omapl1x7()) {		/* or omapl1x7 */
+		ngpio = 128;
 	} else {
 		/* if cpu_is_davinci_dm643x() ngpio = 111 */
 		pr_err("GPIO setup:  how many GPIOs?\n");
@@ -278,6 +280,8 @@ static int __init davinci_gpio_irq_setup
 		bank_irq = IRQ_GPIOBNK0;
 	} else if (cpu_is_davinci_dm646x()) {
 		bank_irq = IRQ_DM646X_GPIOBNK0;
+	} else if (cpu_is_omapl1x7()) {
+		bank_irq = IRQ_OMAPL1X7_GPIO0;
 	} else {
 		printk(KERN_ERR "Don't know first GPIO bank IRQ.\n");
 		return -EINVAL;
Index: primus-kernel/arch/arm/mach-davinci/include/mach/gpio.h
===================================================================
--- primus-kernel.orig/arch/arm/mach-davinci/include/mach/gpio.h
+++ primus-kernel/arch/arm/mach-davinci/include/mach/gpio.h
@@ -18,7 +18,11 @@
 
 #include <mach/irqs.h>
 
+#ifndef CONFIG_ARCH_OMAPL1x7
 #define DAVINCI_GPIO_BASE 0x01C67000
+#else
+#define DAVINCI_GPIO_BASE 0x01E26000
+#endif
 
 /*
  * basic gpio routines
Index: primus-kernel/arch/arm/mach-davinci/omapl1x7.c
===================================================================
--- primus-kernel.orig/arch/arm/mach-davinci/omapl1x7.c
+++ primus-kernel/arch/arm/mach-davinci/omapl1x7.c
@@ -157,6 +157,12 @@ static struct clk emac_clk = {
 	.lpsc = OMAPL1X7_LPSC_CPGMAC,
 };
 
+static struct clk gpio_clk = {
+	.name = "gpio",
+	.parent = &pll1_sysclk4,
+	.lpsc = OMAPL1X7_LPSC_GPIO,
+};
+
 static struct clk *omapl1x7_clks[] __initdata = {
 	&ref_clk,
 	&aux_clkin,
@@ -178,6 +184,7 @@ static struct clk *omapl1x7_clks[] __ini
 	&watchdog_clk,
 	&i2c_clk,
 	&emac_clk,
+	&gpio_clk,
 	NULL,
 };
 
